tool/planet/membership.go

```diff
@@ -0,0 +1,161 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	serf "github.com/hashicorp/serf/client"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// startSerfReconciler creates a control loop that periodically attempts to
+// reconcile serf cluster.
+func startSerfReconciler(ctx context.Context, kubeconfig *rest.Config, serfConfig *serf.Config) {
+	const reconcileInterval = time.Minute * 10
+	ticker := time.NewTicker(reconcileInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := reconcileSerfCluster(kubeconfig, serfConfig); err != nil {
+				log.WithError(err).Warn("Failed to reconcile serf cluster.")
+			}
+		case <-ctx.Done():
+			log.Debug("Stopping serf reconciler")
+			return
+		}
+	}
+}
+
+// reconcileSerfCluster reconciles ther serf cluster if any members are missing
+// from the serf cluster.
+func reconcileSerfCluster(kubeconfig *rest.Config, serfConfig *serf.Config) error {
+	k8sClient, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		return trace.Wrap(err, "failed to create kubernetes clientset")
+	}
+
+	serfClient, err := serf.ClientFromConfig(serfConfig)
+	if err != nil {
+		return trace.Wrap(err, "failed to create serf client")
+	}
+	defer serfClient.Close()
+
+	return reconcileSerf(k8sClient, serfClient)
+}
+
+// reconcileSerf reconciles the serf cluster if any members are missing from the
+// cluster.
+func reconcileSerf(k8sClient kubernetes.Interface, serfClient serfClient) error {
+	serfPeers, err := getSerfPeers(serfClient)
+	if err != nil {
+		return trace.Wrap(err, "failed to get serf peers")
+	}
+
+	k8sPeers, err := getK8sPeers(k8sClient)
+	if err != nil {
+		return trace.Wrap(err, "failed to get k8s peers")
+	}
+
+	missingPeers := getMissing(k8sPeers, serfPeers)
+	if len(missingPeers) == 0 {
+		return nil
+	}
+
+	if _, err := serfClient.Join(missingPeers, false); err != nil {
+		return trace.Wrap(err, "failed to join nodes")
+	}
+
+	log.WithField("prev-cluster", serfPeers).
+		WithField("joined-peers", missingPeers).
+		Info("Successfully reconciled serf cluster.")
+
+	return nil
+}
+
+// getK8sPeers returns advertised IP addresses of all nodes in the kubernetes
+// cluster.
+func getK8sPeers(client kubernetes.Interface) (peers []string, err error) {
+	nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return peers, trace.Wrap(err, "failed to list kubernetes nodes")
+	}
+
+	for _, node := range nodes.Items {
+		addr, exists := node.Labels[advertiseIPKey]
+		if !exists {
+			return peers, trace.NotFound("node %s does not have %s label", node.Name, advertiseIPKey)
```

Comment:
I would just log the warning and skip over the node if it doesn't have this label for some reason and return the rest. Also, how about failing over to `status.addresses` field in case this label doesn't exist? And skipping over the node if that's empty too.